### PR TITLE
Custom background image

### LINKED ISSUE #1059: Define custom background image

### PR DESCRIPTION
**Pull request by 20seven**
Adds setting to define custom background image saved locally
Fixes #1059

### CODE REVIEW

File: src/main/resources/theme/style.css [hunk 1]
```diff
@@ -711,10 +711,7 @@
 /***************** Main window *****************/
 
 .main-window {
-  -fx-background-image: url('images/login-background.jpg');
-  -fx-background-repeat: no-repeat;
-  -fx-background-position: center;
-  -fx-background-size: cover, auto;
+
 }
```
Comment (20seven) [inline src/main/resources/theme/style.css]: Was hesitating if I should have removed the entire class, as well as from styleClass="main-window" in the fxml. I guess I should?

**Review by 1-alex98 — DISMISSED** [review]

File: src/main/java/com/faforever/client/main/MainController.java [hunk 2]
```diff
@@ -329,7 +336,7 @@ public void display() {
     eventBus.post(UpdateApplicationBadgeEvent.ofNewValue(0));
 
     Stage stage = StageHolder.getStage();
-
+    setBackgroundImage(preferencesService.getPreferences().getMainWindow().backgroundImagePathProperty().getValue());
```
Comment (micheljung) [inline src/main/java/com/faforever/client/main/MainController.java]:
```suggestion
    setBackgroundImage(preferencesService.getPreferences().getMainWindow().getBackgroundImagePath());
```

File: src/main/java/com/faforever/client/main/MainController.java [hunk 3]
```diff
@@ -422,6 +429,27 @@ private void registerWindowListeners() {
         preferencesService.storeInBackground();
       }
     });
+    JavaFxUtil.addListener(mainWindowPrefs.backgroundImagePathProperty(), observable -> {
+      setBackgroundImage(mainWindowPrefs.backgroundImagePathProperty().getValue());
```
Comment (micheljung) [inline src/main/java/com/faforever/client/main/MainController.java]:
```suggestion
      setBackgroundImage(mainWindowPrefs.getBackgroundImagePath());
```

File: src/main/java/com/faforever/client/main/MainController.java [hunk 4]
```diff
@@ -422,6 +429,27 @@ private void registerWindowListeners() {
...
+    });
+  }
+
+  private void setBackgroundImage(String filepath) {
```
Comment (micheljung) [inline src/main/java/com/faforever/client/main/MainController.java]:
```suggestion
  private void setBackgroundImage(Path filepath) {
```

File: src/main/java/com/faforever/client/main/MainController.java [hunk 5]
```diff
@@ -422,6 +429,27 @@ private void registerWindowListeners() {
...
+    Image image;
+    if (filepath == null || filepath.isEmpty()) {
+      image = new Image(uiService.getThemeFile("theme/images/login-background.jpg"));
+    } else {
+      File file = new File(filepath);
```
Comment (micheljung) [inline src/main/java/com/faforever/client/main/MainController.java]: `java.io.File` is pre-Java 7 and should be avoided, instead use `java.nio.file.Path`

File: src/main/java/com/faforever/client/preferences/WindowPrefs.java [hunk 6]
```diff
@@ -22,6 +22,7 @@
   private final MapProperty<String, String> lastChildViews;
   private final DoubleProperty x;
   private final DoubleProperty y;
+  private final StringProperty backgroundImagePath;
```
Comment (micheljung) [inline src/main/java/com/faforever/client/preferences/WindowPrefs.java]:
```suggestion
  private final ObjectProperty<Path> backgroundImagePath;
```

File: src/main/java/com/faforever/client/preferences/WindowPrefs.java [hunk 7]
```diff
@@ -31,6 +32,7 @@ public WindowPrefs() {
     maximized = new SimpleBooleanProperty();
     lastView = new SimpleStringProperty();
     lastChildViews = new SimpleMapProperty<>(FXCollections.observableHashMap());
+    backgroundImagePath = new SimpleStringProperty();
```
Comment (micheljung) [inline src/main/java/com/faforever/client/preferences/WindowPrefs.java]:
```suggestion
    backgroundImagePath = new SimpleObjectProperty<>();
```

File: src/main/java/com/faforever/client/preferences/ui/SettingsController.java [hunk 8]
```diff
@@ -442,5 +448,20 @@ public void onPreviewToastButtonClicked() {
...
+  public void onSelectBackgroundImage() {
+    Platform.runLater(() -> {
+      FileChooser directoryChooser = new FileChooser();
+      directoryChooser.setTitle(i18n.get("settings.appearance.chooseImage"));
+      File result = directoryChooser.showOpenDialog(getRoot().getScene().getWindow());
+
+      if (result == null) {
+        return;
+      }
+      preferencesService.getPreferences().getMainWindow().setBackgroundImagePath(result.getPath());
```
Comment (micheljung) [inline src/main/java/com/faforever/client/preferences/ui/SettingsController.java]:
```suggestion
      preferencesService.getPreferences().getMainWindow().setBackgroundImagePath(result.toPath());
```

File: src/main/resources/theme/style.css (hunk 1 above)
Comment (micheljung) [reply src/main/resources/theme/style.css]: It's fine if you keep it in the FXML, allows theme creators to modify it. In CSS, you can remove it.

**Review by micheljung — CHANGES_REQUESTED** [review]

**Review by 1-alex98 — DISMISSED** [review]

**Pull request merged**